blockchain/thresholdstate.go

```diff
@@ -465,6 +465,82 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
 	return invalidState, DeploymentError(deploymentID)
 }
 
+// thresholdStatelastChanged returns the last node before the current state of
+// consensus deployment agenda activated.
+func (b *BlockChain) thresholdStateLastChanged(version uint32, prevNode *blockNode, checker thresholdConditionChecker, cache *thresholdStateCache) (*blockNode, error) {
+	// The threshold state for the window that contains the genesis block is
+	// defined by definition.
+	confirmationWindow := int64(checker.RuleChangeActivationInterval())
+	svh := checker.StakeValidationHeight()
+	if prevNode == nil || prevNode.height+1 < svh+confirmationWindow {
+		return b.bestChain.Genesis(), nil
+	}
+
+	curState, err := b.thresholdState(version, prevNode, checker, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the ancestor that is the last block of the previous confirmation
+	// window in order to get its threshold state.  This can be done because
+	// the state is the same for all blocks within a given window.
+	wantHeight := calcWantHeight(svh, confirmationWindow, prevNode.height+1)
+	prevNode = prevNode.Ancestor(wantHeight)
+	for prevNode != nil {
+		state, err := b.thresholdState(version, prevNode, checker, cache)
+		if err != nil {
+			return nil, err
+		}
+
+		if state.State != curState.State {
+			return prevNode, nil
+		}
+
+		// Get the ancestor that is the last block of the previous
+		// confirmation window.
+		prevNode = prevNode.RelativeAncestor(confirmationWindow)
+	}
+
+	return b.bestChain.Genesis(), nil
+}
+
+// StateActivationHeight returns the height of the first block where the
+// current state of consensus deployment agenda activated.
+func (b *BlockChain) StateActivationHeight(hash *chainhash.Hash, version uint32, deploymentID string) (int64, error) {
+	// Fetch the block node of the provided hash.
+	node := b.index.LookupNode(hash)
+	if node == nil || !b.index.NodeStatus(node).KnownValid() {
+		return 0, HashError(hash.String())
```

Comment:
Not the correct error.  It should be the same as the others which involve lookups.  Also, the node status is irrelevant here.